### ISSUE TITLE
about tree-shaking

### ISSUE THREAD

**Issue by wdssmq** [rollup/rollup#4488]
```js
// base.js
const str1 = "1111";

const str2 = "2222".trim();

const str3 = `3333`;

const str4 = `4444`.trim();

const $ = window.$ || unsafeWindow.$;

export { str1, str2, str3, str4, $ };
```

```js
// main.js
import { str1, str2, str3, str4, $ } from "./base";

console.log(str1);
```
bundle：
```js
(function () {
	'use strict';

	// base.js
	const str1 = "1111";

`4444`	.trim();

	window.$ || unsafeWindow.$;

	// main.js

	console.log(str1);

})();
```

如何不输出「`4444`	.trim();」和 `window.$ || unsafeWindow.$;`两行？

How to not output 「`4444`	.trim();」and `window.$ || unsafeWindow.$;`?

**Comment by dnalborczyk** [comment]
@wdssmq are you using an old version of rollup?

https://rollupjs.org/repl/?version=2.72.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwc3RyMSUyMCUzRCUyMCU1QyUyMjExMTElNUMlMjIlM0IlNUNuJTVDbmNvbnN0JTIwc3RyMiUyMCUzRCUyMCU1QyUyMjIyMjIlNUMlMjIudHJpbSgpJTNCJTVDbiU1Q25jb25zdCUyMHN0cjMlMjAlM0QlMjAlNjAzMzMzJTYwJTNCJTVDbiU1Q25jb25zdCUyMHN0cjQlMjAlM0QlMjAlNjA0NDQ0JTYwLnRyaW0oKSUzQiU1Q24lNUNuY29uc3QlMjAlMjQlMjAlM0QlMjB3aW5kb3cuJTI0JTIwJTdDJTdDJTIwdW5zYWZlV2luZG93LiUyNCUzQiU1Q24lNUNuZXhwb3J0JTIwJTdCJTIwc3RyMSUyQyUyMHN0cjIlMkMlMjBzdHIzJTJDJTIwc3RyNCUyQyUyMCUyNCUyMCU3RCUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

**Comment by dnalborczyk** [comment]
nm, your question is probably related to:

https://rollupjs.org/repl/?version=2.72.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMGJhc2UuanMlNUNuY29uc3QlMjBzdHIxJTIwJTNEJTIwJTVDJTIyMTExMSU1QyUyMiUzQiU1Q24lNUNuY29uc3QlMjBzdHIyJTIwJTNEJTIwJTVDJTIyMjIyMiU1QyUyMi50cmltKCklM0IlNUNuJTVDbmNvbnN0JTIwc3RyMyUyMCUzRCUyMCU2MDMzMzMlNjAlM0IlNUNuJTVDbmNvbnN0JTIwc3RyNCUyMCUzRCUyMCU2MDQ0NDQlNjAudHJpbSgpJTNCJTVDbiU1Q25jb25zdCUyMCUyNCUyMCUzRCUyMHdpbmRvdy4lMjQlMjAlN0MlN0MlMjB1bnNhZmVXaW5kb3cuJTI0JTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

**Comment by dnalborczyk** [comment]
I'll let others chime in, but this ```const str4 = `4444`.trim();``` appears to be a bug, or a feature.

the second `const $ = window.$ || unsafeWindow.$;`  looks correct as `window.$` and the global `unsafeWindow.$` could have side potential effects.

that said, I think I still see a bug, unless it's by design:

I believe this alone should not be tree-shaken: `const $ = window.$`, as a $ getter could have side effects.

https://rollupjs.org/repl/?version=2.72.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwJTI0JTIwJTNEJTIwd2luZG93LiUyNCUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

**Comment by wdssmq** [comment]
@dnalborczyk 
Thanks for the reply.

Some of the older monkey scripts use jQuery, which I will gradually replace with native js, or comment out.

The other issue does look like a bug, and may need further confirmation.

**Comment by lukastaegert** [comment]
As for the `window` question, Rollup does not know the `$` global and more importantly does not know the `unsafeWindow` global. Accessing a non-existing global will throw an error, so Rollup needs to keep the code. Setting `treeshake.unknownGlobalSideEffects` to `false` will probably fix this, but maybe you also need `treeshake.propertyReadSideEffects`, see https://rollupjs.org/guide/en/#treeshake.

As for retaining `.trim()`, this will be fixed by #4511
